add metric for time retention failing due to end time

jadami10 opened the pull request:
This is a new metric to catch when segment start/end time is invalid and there time based segment retention is failing.

satwik-pachigolla reviewed (approved):
Looks good to me!

vvivekiyer commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -1505,7 +1505,7 @@ public void uploadToDeepStoreIfMissing(TableConfig tableConfig, List<SegmentZKMe
           .toMillis(Long.parseLong(validationConfig.getRetentionTimeValue()));
     RetentionStrategy retentionStrategy = new TimeRetentionStrategy(TimeUnit.MILLISECONDS,
-          retentionMs - MIN_TIME_BEFORE_SEGMENT_EXPIRATION_FOR_FIXING_DEEP_STORE_COPY_MILLIS);
+        retentionMs - MIN_TIME_BEFORE_SEGMENT_EXPIRATION_FOR_FIXING_DEEP_STORE_COPY_MILLIS, _controllerMetrics);
```
Is it desirable behavior to update this metric when isPurgeable returns false when uploading the LLC segment to deepstore?

jadami10 replied:
this metric isn't for all instances of isPurgeable=false. It's specifically the case where the zk end time is in an unexpected format, so isPurgeable will always be false. Rather than "this segment just isn't old enough yet".

tibrewalpratik17 replied:
Can we instead have this metric emission from `SegmentStatusChecker#updateSegmentMetrics`? We already fetch segmentZKMetadata over there as well and can easily validate this field.

We can also create a new periodic task - `SegmentZKMetadataValidator` and easily extend it to have multiple validations in future. We can keep retention-flow separate of this metadata-validation imo.

jadami10 replied:
The reason I chose not to do it async is because this is the place that's doing the pruning check and failing. If we do it In `SegmentStatusChecker`, then we need to take into account "does this table even have retention enabled". Because that's the only time, as far as I know, that we really care about this case.

I guess I'm looking at this less like a "segment checker" and more like "the retention feature is failing"

tibrewalpratik17 replied:
Having this check in "segment checker" covers "retention feature is failing" plus gives us the scope of extending it to have multiple different checks in future.

jadami10 replied:
not exactly. We'd have to duplicate the logic in https://github.com/apache/pinot/blob/master/pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/retention/RetentionManager.java#L95 into the status checker as well. Or make "isRetentionEnabled" a static function and call it.

Either way, seems like we disagree on where to place this. Let's get a third opinion. cc @Jackie-Jiang, curious on your thoughts since you've most recently touched both the `SegmentStatusChecker` and `TimeRetentionStrategy` code

Jackie-Jiang reviewed (commented):
We do have a check in `TimeValidationTransformer` to check if the time value is in valid range. Can you double check why it is not triggered since you don't have `continueOnError` set to true?
It is good to emit a metric on segments with invalid time value. I feel `SegmentStatusChecker` might be a better place for such check

jadami10 commented:
> We do have a check in `TimeValidationTransformer` to check if the time value is in valid range. Can you double check why it is not triggered since you don't have `continueOnError` set to true?

Because there's no `ingestionConfig` set, we hit [this branch](https://github.com/apache/pinot/blob/master/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/recordtransformer/TimeValidationTransformer.java#L60) that sets `_enableTimeValueCheck` to false.

But this code also has a similar problem. https://github.com/apache/pinot/blob/master/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/recordtransformer/TimeValidationTransformer.java#L93 sets the time value to null without incrementing any metrics. Pinot is basically failing to do something quite important (ingest the data) without offering any way to alert users.

jadami10 commented:
I still don't agree that SegmentStatusChecker is the right place for this, but I've added all "invalid time" metrics to it. I didn't duplicate the "is retention-able" logic, because that would add a 3rd this is duplicated. I've added a unit test on the metric as well.

Jackie-Jiang commented:
Good point on emitting metrics when error is captured. The reason why we didn't emit metrics before is because there was no way to pass in a metric object into the `RecordTransformer` as it can be running in different components (e.g. Server/Minion/Hadoop/Spark). Since we have added singleton metric object, we can emit metrics now.

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/common/metrics/ControllerGauge.java:
```diff
@@ -35,6 +35,11 @@ public enum ControllerGauge implements AbstractMetrics.Gauge {
+  // Segment start and end time is stored in milliseconds.
+  // Invalid start/end time means the broker time pruner will not work correctly.
+  // Invalid end times means time retention will not happen for that segment.
+  SEGMENT_INVALID_START_TIME("segmentInvalidStartTime", false),
+  SEGMENT_INVALID_END_TIME("segmentInvalidEndTime", false),
```
```suggestion
  SEGMENTS_WITH_INVALID_START_TIME("segments", false),
  SEGMENTS_WITH_INVALID_END_TIME("segments", false),
```

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/SegmentStatusChecker.java:
```diff
@@ -318,6 +321,17 @@ private void updateSegmentMetrics(String tableNameWithType, TableConfig tableCon
         continue;
       }
 
+      long startTimeMs = segmentZKMetadata.getStartTimeMs();
+      boolean isStartTimeMsInvalid = startTimeMs > 0 && !TimeUtils.timeValueInValidRange(startTimeMs);
+      long endTimeMs = segmentZKMetadata.getEndTimeMs();
+      boolean isEndTimeMsInvalid = endTimeMs > 0 && !TimeUtils.timeValueInValidRange(endTimeMs);
+      if (isStartTimeMsInvalid) {
+        segmentsInvalidStartTime.add(segment);
+      }
+      if (isEndTimeMsInvalid) {
+        segmentsInvalidEndTime.add(segment);
+      }
```
Let's only skip the `IN_PROGRESS` segments and count negative start/end time as invalid:
```
    ...
  }
```
```suggestion
      if (segmentZKMetadata.getStatus() != Status.IN_PROGRESS) {
        if (!TimeUtils.timeValueInValidRange(segmentZKMetadata.getStartTimeMs()) {
          segmentsInvalidStartTime.add(segment);
        }
        if (!TimeUtils.timeValueInValidRange(segmentZKMetadata.getEndTimeMs()) {
          segmentsInvalidEndTime.add(segment);
        }
      }
```

Jackie-Jiang reviewed (approved):
LGTM otherwise
